sqlanalytics/resource_widget.go

```diff
@@ -0,0 +1,312 @@
+			SizeX: pos.SizeX,
+			SizeY: pos.SizeY,
+			PosX:  pos.PosX,
+			PosY:  pos.PosY,
+		}
+	}
+
+	w.Parameter = make([]WidgetParameter, 0, len(aw.Options.ParameterMapping))
+	for _, p := range aw.Options.ParameterMapping {
+		wp := WidgetParameter{
+			Name:  p.Name,
+			Type:  p.Type,
+			MapTo: p.MapTo,
+			Title: p.Title,
+		}
+
+		// Re-marshal value so we can try to unmarshal different types.
+		// We don't know about the type it holds, because it depends
+		// on the parameter's type, which we don't have access to.
+		b, err := json.Marshal(p.Value)
+		if err != nil {
+			return err
+		}
+
+		// Try unmarshalling `string`.
+		{
+			var v string
+			err := json.Unmarshal(b, &v)
+			if err == nil {
+				wp.Value = v
+				w.Parameter = append(w.Parameter, wp)
+				continue
+			}
+		}
+
+		// Try unmarshalling `[]string`.
+		{
+			var vs []string
+			err := json.Unmarshal(b, &vs)
+			if err == nil {
+				wp.Values = vs
+				w.Parameter = append(w.Parameter, wp)
+				continue
+			}
+		}
+
+		return fmt.Errorf("Unable to derive type from message: %v", string(b))
+	}
+
+	// Sort parameters by their name for deterministic order.
+	sort.Sort(sortWidgetParameter(w.Parameter))
+
+	// Pass to ResourceData.
+	return common.StructToData(*w, schema, data)
+}
+
+// NewWidgetAPI ...
+func NewWidgetAPI(ctx context.Context, m interface{}) WidgetAPI {
+	return WidgetAPI{m.(*common.DatabricksClient), ctx}
+}
+
+// WidgetAPI ...
+type WidgetAPI struct {
+	client  *common.DatabricksClient
+	context context.Context
+}
+
+// Create ...
+func (a WidgetAPI) Create(w *api.Widget) error {
+	return a.client.Post(a.context, "/preview/sql/widgets", w, &w)
+}
+
+// Read ...
+func (a WidgetAPI) Read(dashboardID, widgetID string) (*api.Widget, error) {
+	d, err := NewDashboardAPI(a.context, a.client).Read(dashboardID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look for matching widget ID.
+	for _, wp := range d.Widgets {
+		var wnew api.Widget
+		err = json.Unmarshal(wp, &wnew)
+		if err != nil {
+			return nil, err
+		}
+
+		if strconv.Itoa(wnew.ID) == widgetID {
+			// Include dashboard ID in returned object.
+			// It's not part of the API response.
+			wnew.DashboardID = dashboardID
+			return &wnew, nil
+		}
+	}
+
+	return nil, fmt.Errorf("Cannot find widget %s attached to dashboard %s", widgetID, dashboardID)
+}
+
+// Update ...
+func (a WidgetAPI) Update(widgetID string, w *api.Widget) error {
+	return a.client.Post(a.context, fmt.Sprintf("/preview/sql/widgets/%s", widgetID), w, nil)
+}
+
+// Delete ...
+func (a WidgetAPI) Delete(widgetID string) error {
+	return a.client.Delete(a.context, fmt.Sprintf("/preview/sql/widgets/%s", widgetID), nil)
+}
+
+// ResourceWidget ...
+func ResourceWidget() *schema.Resource {
+	p := common.NewPairSeparatedID("dashboard_id", "widget_id", "/")
+	s := common.StructToSchema(
+		WidgetEntity{},
+		func(m map[string]*schema.Schema) map[string]*schema.Schema {
+			m["text"].ConflictsWith = []string{"visualization_id"}
```

Comment:
don't forget to add ForceNew to dashboard_id & widget_id

Reply:
Done for both widget and visualization (both use composite ID).